controllers/helmrelease_controller_release.go

```diff
@@ -0,0 +1,232 @@
+/*
+Copyright 2020 The Flux CD contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/fluxcd/pkg/apis/meta"
+	"github.com/fluxcd/pkg/runtime/events"
+	"k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	v2 "github.com/fluxcd/helm-controller/api/v2beta1"
+	"github.com/fluxcd/helm-controller/internal/runner"
+	"github.com/fluxcd/helm-controller/internal/util"
+)
+
+func (r *HelmReleaseReconciler) reconcileRelease(ctx context.Context, run *runner.Runner, hr *v2.HelmRelease) (ctrl.Result, error) {
+	// Get the Helm chart
+	hc, err := r.getHelmChart(ctx, hr)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Check chart readiness
+	if hc.Generation != hc.Status.ObservedGeneration || !meta.HasReadyCondition(hc.Status.Conditions) {
+		err := fmt.Errorf("HelmChart '%s/%s' is not ready", hc.GetNamespace(), hc.GetName())
+		return ctrl.Result{}, err
+	}
+
+	// Compose the values for this release, and record the checksum
+	values, err := r.composeValues(ctx, *hr)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	valuesChecksum := util.ValuesChecksum(values)
+
+	// Retrieve the last release for observation
+	rel, err := run.ObserveLastRelease(*hr)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// If new state, mark release as progressing which resets the conditions
+	if hr.Status.LastAttemptedRevision != hc.GetArtifact().Revision ||
+		hr.Status.LastAttemptedValuesChecksum != valuesChecksum ||
+		hr.Status.LastReleaseRevision != util.ReleaseRevision(rel) ||
+		hr.Status.ObservedGeneration != hr.Generation {
+		v2.HelmReleaseProgressing(hr)
+		if err := r.Client.Status().Update(ctx, hr); err != nil {
+			return ctrl.Result{}, err
+		}
+		r.recordReadiness(hr, false)
+	}
+
+	// If the previous release was successful, return
+	if released := meta.GetCondition(hr.Status.Conditions, v2.ReleasedCondition); released != nil {
+		if released.Status == v1.ConditionTrue {
+			return ctrl.Result{}, nil
```

Comment:
We should "remember" the previous Ready reason/message here, which is stored in the Released condition, since it may have been lost in e.g. a transient dependency not ready reconciliation.